### PR TITLE
Panache REST guide

### PR DESCRIPTION
**Pull request by gytis**
cc @geoand

### CODE REVIEW

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [inline docs/src/main/asciidoc/panache-rest.adoc:131]: It's out of scope for this PR but I'm not happy with a name of this annotation (`@PanacheRestResource`). It's very similar to the interface name (`PanacheCrud..Resource`) and might be confusing. Any suggestions on the name change @geoand?

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [inline docs/src/main/asciidoc/panache-rest.adoc:12]:
```suggestion
A lot of web applications are monotonous CRUD applications with REST APIs that are tedious to write.
```

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [inline docs/src/main/asciidoc/panache-rest.adoc:13]:
```suggestion
To streamline this task, Panache REST can generate the basic CRUD endpoints for your entities and repositories.
```

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [inline docs/src/main/asciidoc/panache-rest.adoc:15]: Perhaps this should be in a note?

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [inline docs/src/main/asciidoc/panache-rest.adoc:59]:
```suggestion
_Do not implement these interfaces and don't provide custom methods because they be ignored._
```

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [inline docs/src/main/asciidoc/panache-rest.adoc:63]: This line doesn't seem to make sense

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [inline docs/src/main/asciidoc/panache-rest.adoc:66]: I think we need a little more emphasis that on the fact this is generated - not something the user would write.
Mentioning it because looking at the code initially through me off...

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [inline docs/src/main/asciidoc/panache-rest.adoc:111]:
```suggestion
If your application has an entity (e.g. `Person`) that extends either `PanacheEntity` or `PanacheEntityBase` class, you could instruct Panache REST to generate its CRUD resource with the following interface:
```

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [inline docs/src/main/asciidoc/panache-rest.adoc:121]:
```suggestion
If your application has a simple entity (e.g. `Person`) and a repository (e.g. `PersonRepository`) that implements either `PanacheRepository` or `PanacheRepositoryBase` interface, you could instruct Panache REST to generate its CRUD resource with the following interface:
```

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [inline docs/src/main/asciidoc/panache-rest.adoc:145]: Perhaps a link to what `hal` is would be nice

**Review by geoand — COMMENTED** [review]
Nice start!

Added some comments

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [reply docs/src/main/asciidoc/panache-rest.adoc:131]: I dunno.... I'm probably the worst developer when it comes to naming...

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:15]: There is a generic note about the experimental feature already. Do you think I should add another one?

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:59]: `will be ignored` maybe?

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [reply docs/src/main/asciidoc/panache-rest.adoc:15]: Oh OK. If it's used elsewhere as well, let's keep it as it is

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [reply docs/src/main/asciidoc/panache-rest.adoc:59]: yeah :)

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:63]: How about `The generated resources will be functionally equivalent for both entities and repositories. Except that the particular data access pattern will be used.`

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:66]: It does say "generated JAX-RS resource". Any suggestion?

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:145]: I thought about it, but then though it might be too much info. Do you think it would be useful?

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [reply docs/src/main/asciidoc/panache-rest.adoc:63]: My counter: 

```
The generated resources will be functionally equivalent for both entities and repositories., the difference being in the particular data access pattern will be used.
```

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [reply docs/src/main/asciidoc/panache-rest.adoc:66]: Yeah I know, it's just that it's too easy to think you need to write this. 
Perhaps change the name of the class and add a comment on it?

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [reply docs/src/main/asciidoc/panache-rest.adoc:145]: I think it would be nice to have

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:59]: How do you edit string in the PR?

Comment (geoand) [reply docs/src/main/asciidoc/panache-rest.adoc:59]: See `6` from https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:63]: `The only difference being the particular data access pattern in use.`?

Comment (geoand) [reply docs/src/main/asciidoc/panache-rest.adoc:63]: Yeah sorry. I really need to pay more attention to what I actually write as opposed to what I am thinking...

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:145]: Done

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:66]: I've changed the sentence above

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [inline docs/src/main/asciidoc/panache-rest.adoc:27]:
```suggestion
** One of the RESTEasy JSON serialization extensions (`quarkus-resteasy-jackson` or `quarkus-resteasy-jsonb`)
```

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/panache-rest.adoc:6]: Please avoid the newline here, the website is very picky about that.

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/panache-rest.adoc:58]:
```suggestion
For each entity and repository you want to expose as a REST service, provide an interface that extends either `PanacheEntityCrudResource` or `PanacheRepositoryCrudResource` interface.
```
would be more correct if I understand the extension correctly?

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/panache-rest.adoc:68]: I think this is a mistake. We shouldn't "pluralize" things as there's a good chance we will get some of them wrong and it really doesn't bring anything.

Keep in mind that a lot of people code in their own languages where the plural rules will be completely different.

Also, frankly, I don't think it's better as a plural so why take the risk?

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/panache-rest.adoc:145]: Wondering if we should have an array of content types instead with the default being `{ "application/json" }`. That would be much more flexible.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added some comments, some are related to the implementation itself rather than the doc.

/cc @geoand for further discussions.

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gsmet) [reply docs/src/main/asciidoc/panache-rest.adoc:68]: /cc @FroMage

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gsmet) [reply docs/src/main/asciidoc/panache-rest.adoc:145]: /cc @FroMage

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [reply docs/src/main/asciidoc/panache-rest.adoc:68]: If we want plural, perhaps let the user specify it?

Comment (gsmet) [reply docs/src/main/asciidoc/panache-rest.adoc:68]: Just to give an example, in French, we have a lot of specific plural forms (`cheval` -> `chevaux` for instance).

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (geoand) [reply docs/src/main/asciidoc/panache-rest.adoc:145]: Well, I don't think we currently support any response type, but this seems like a valid point for the future.

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gsmet) [reply docs/src/main/asciidoc/panache-rest.adoc:68]: Yeah, taking the entity name as is would be the default and you could define a specific path in the annotation?

Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:68]: Would you mind to create an issue for this and assign to me? I'll need to change the extension because it now pluralises everything.
I'll also add a way to override it separately.

Just an FIY Spring Data REST simply adds `s` to every entity name.

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:145]: `application/json` is hardcoded and `application/hal+json` can be enabled as an extra. Because HAL is a complex format I had to implement a custom serialiser for it which serialisers special wrapper objects.

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gsmet) [reply docs/src/main/asciidoc/panache-rest.adoc:68]: > Just an FIY Spring Data REST simply adds s to every entity name.

Oh, my. OK I will create an issue.

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gsmet) [reply docs/src/main/asciidoc/panache-rest.adoc:145]: OK, let's keep it that way for now.

**Comment by gytis** [comment]
I think I've addresses all the comments except the one about plural paths. For that I'll make a separate guide modification once I change the extension code.

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (FroMage) [reply docs/src/main/asciidoc/panache-rest.adoc:67]: Makes it even more fun with:

```java
@Entity
public class 馬 extends PanacheEntity {
}
```

And the URI ends up being `/馬s`, right? :)

Myself, I've always used `Cheval` for the entity name (singular) and `Chevaux` (plural) for the controller name. And RoR derives the URIs from the controller names, so it would be `/chevaux` correctly.

I just suggest we advertise the same convention, it solves many issues. And we default to the controller name for the URI, instead of the entity name.

Naturally we should also allow users to override it.

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/panache-rest.adoc:51]:
```suggestion
* Implement the Panache entities and/or repositories as explained in the link:hibernate-orm-panache[Hibernate ORM with Panache guide].
```

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:67]: > And the URI ends up being `/馬s`, right? :)

I would think so. At least for a simple `people` I got a path `/peoples` when i tried it last time.

> I just suggest we advertise the same convention, it solves many issues. And we default to the controller name for the URI, instead of the entity name.

Extracting a name from a controller will be trickier because a user can name his classes differently. We could recommend a convention but who knows if they'll follow it.
With an entity we could simply take an its name, make it all lowercase separated with `-` instead of camelcase and get on with it. And then I'll add an option to override the path.

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/panache-rest.adoc:24]: Now that I read it like that, I've my doubts over the order, because originally it was supposed to be `X with Panache`, which means `X with style`.

This makes it looks like `Panache` is a noun and the subject instead of an adjective to make Hibernate ORM/REST stylish.

Perhaps it should be `Hibernate ORM REST with Panache` or `RESTful Hibernate ORM with Panache` or `REST for Hibernate ORM with Panache`?

Comment (FroMage) [reply docs/src/main/asciidoc/panache-rest.adoc:24]: I guess, similarly, this guide (and the extension names) would better be "REST with Panache" and `rest-panache`?

Except this sort-of implies it applies to REST apps that aren't necessarily about data, and we plan to have a Panache layer for RESTEasy one day…

Perhaps `RESTful data with Panache`?

Comment (FroMage) [reply docs/src/main/asciidoc/panache-rest.adoc:24]: Or `CRUD with Panache`?

**Review by FroMage — CHANGES_REQUESTED** [review]
Great guide. Please don't hate me for reconsidering naming before it's too late…

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (FroMage) [reply docs/src/main/asciidoc/panache-rest.adoc:67]: > Extracting a name from a controller will be trickier because a user can name his classes differently. We could recommend a convention but who knows if they'll follow it.

The point of conventions is that people follow them. If they don't they can't customise the output.

Entities are already singular by convention. Controller names don't have a convention other than the brain-dead "resource" that jax-rs got confused about.

It's pretty easy to tell people to name their controller `People` and default to using that name.

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gsmet) [reply docs/src/main/asciidoc/panache-rest.adoc:24]: In any case, we need to make the extension name consistent with the naming rules we define here.

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:67]: Let's discuss it on https://github.com/quarkusio/quarkus/issues/9261 since the guide change won't be enough for this anyway.

File: docs/src/main/asciidoc/panache-rest.adoc
Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:24]: Should I create a separate issue to agree on the name there?

Comment (FroMage) [reply docs/src/main/asciidoc/panache-rest.adoc:24]: We can, yes, but this will heavily impact this guide.

Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:24]: I could just update the guide together with the code once we agree on a name.

Comment (FroMage) [reply docs/src/main/asciidoc/panache-rest.adoc:24]: OK fine.

Comment (gytis) [reply docs/src/main/asciidoc/panache-rest.adoc:24]: https://github.com/quarkusio/quarkus/issues/9293

**Comment by gytis** [comment]
I've updated the guide to align with the changes from https://github.com/quarkusio/quarkus/pull/9373 but I'll need to rebase after the merge. There will be a conflict in a `quarkus-extension.yaml` file because of a module rename.

**Comment by geoand** [comment]
We'll need a review from @FroMage for this one :)

**Comment by gytis** [comment]
Closing this PR because I've cherry-picked the commit to https://github.com/quarkusio/quarkus/pull/9373

**Pull request closed without merging**
